src/object.c

```diff
@@ -211,6 +211,303 @@ robj *dupStringObject(const robj *o) {
+        newis->length = is->length;
+        zfree(set->ptr);
+        set->ptr = newis;
+    } else if (set->encoding == OBJ_ENCODING_HT) {
+        si = setTypeInitIterator(o);
+        while (setTypeNext(si, &elesds, &intobj) != -1) {
+            setTypeAdd(set, elesds);
+        }
+        setTypeReleaseIterator(si);
+    } else {
+        serverPanic("Unknown set encoding");
+    }
+    return set;
+}
+
+/* This is a helper function for the COPY command.
+ * Duplicate a sorted set object, with the guarantee that the returned object
+ * has the same encoding as the original one.
+ * 
+ * The resulting object always has refcount set to 1 */
+robj *dupZsetObject(robj *o) {
+    robj *zobj;
+
+    serverAssert(o->type == OBJ_ZSET);
+    
+    /* Create a new sorted set object that have the same encoding as the original object's encoding */
+    switch (o->encoding) {
+        case OBJ_ENCODING_ZIPLIST:
+            zobj = createZsetZiplistObject();
+            break;
+        case OBJ_ENCODING_SKIPLIST:
+            zobj = createZsetObject();
+            break;
+        default:
+            serverPanic("Wrong encoding.");
+            break;
+    }
+    if (zobj->encoding == OBJ_ENCODING_ZIPLIST) {
+        unsigned char *zl = o->ptr;
+        size_t sz = ziplistBlobLen(zl);
+        unsigned char *new_zl = zmalloc(sz);
+        memcpy(new_zl, zl, sz);
+        zfree(zobj->ptr);
+        zobj->ptr = new_zl;
+    } else if (zobj->encoding == OBJ_ENCODING_SKIPLIST) {
+        zset *zs = o->ptr;
+        zskiplist *zsl = zs->zsl;
+        zskiplistNode *ln;
+        sds ele;
+        int retflags = ZADD_NONE;
+        long llen = zsetLength(o);
+        
+        /* We copy the skiplist elements from the greatest to the
+         * smallest (that's trivial since the elements are already ordered in
+         * the skiplist): this improves the load process, since the next loaded
+         * element will always be the smaller, so adding to the skiplist
+         * will always immediately stop at the head, making the insertion
+         * O(1) instead of O(log(N)). */
+        ln = zsl->tail;
+        while (llen--) {
+            ele = ln->ele;
+            zsetAdd(zobj, ln->score, ele, &retflags, NULL);
+            ln = ln->backward;
+        }
+    } else {
+        serverPanic("Unknown sorted set encoding");
+    }
+    return zobj;
+}
+
+/* This is a helper function for the COPY command.
+ * Duplicate a hash object, with the guarantee that the returned object
+ * has the same encoding as the original one.
+ * 
+ * The resulting object always has refcount set to 1 */
+robj *dupHashObject(robj *o) {
+    robj *hobj;
+    hashTypeIterator *hi;
+
+    serverAssert(o->type == OBJ_HASH);
+
+    switch (o->encoding) {
+        case OBJ_ENCODING_ZIPLIST:
+            hobj = createHashObject();
+            break;
+        case OBJ_ENCODING_HT:
+            hobj = createHashObject();
+            hashTypeConvert(hobj, OBJ_ENCODING_HT);
```

Comment:
it's a good idea to call dictExpand here too, so that it'll be created with the right size and not go though rehashing.
same in the other paces that we create a dict which we know it's final size

Reply:
use `dictSize(d)` instead of `d->ht[0].size`. it's both nicer not to access the internals of the dict, but also more "correct" since if the source dict is in the process of rehashing, it's records are split between ht[0] and ht[1].